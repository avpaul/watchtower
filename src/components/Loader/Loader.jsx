import React from 'react';

import './Loader.css';

/**
 * Loader UI Component
 *
 * @returns {JSX} React component
 */
const Loader = () => (
  <div className="loader__center">
    <div className="loader" />
  </div>
);

export default Loader;
