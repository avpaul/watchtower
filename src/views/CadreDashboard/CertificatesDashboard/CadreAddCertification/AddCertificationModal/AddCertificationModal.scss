.generic-modal textarea {
  height: 9rem;
  resize: none;
}

#saveCertificationButton,
#closeCertificationModal {
  background-color: #10a36d;
  border: #10a36d;
  border-radius: 0.3rem;
  font-weight: bold;
  font-size: 90%;
}

#addCertificationModal .modal-header {
  padding-bottom: 0;
}

#addCertificationModal .modal-footer {
  padding-top: 0;
}

#addCertificationModal h1 {
  font-size: 150%;
  font-weight: 400 !important;
  color: black;
  margin-bottom: 0.3rem;
}

.modal-lg {
  width: 37rem;
}

@media (max-width: 554px) {
  .modal-lg {
    width: 90%;
    margin: 0 auto;
  }
}
