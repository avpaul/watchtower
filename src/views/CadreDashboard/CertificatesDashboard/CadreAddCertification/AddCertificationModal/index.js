import AddCertificationModalContainer from './CertificationsContainer';

export default AddCertificationModalContainer;
