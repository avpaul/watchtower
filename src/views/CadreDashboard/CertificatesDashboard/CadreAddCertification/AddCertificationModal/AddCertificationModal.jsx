import React, { Component } from 'react';
import PropTypes from 'prop-types';
import FormInputs from '../../../../../components/FormInputs';
import GenericModal from '../../../../../components/GenericModal';
import Loader from '../../../../../components/Loader/Loader';
import './AddCertificationModal.scss';
import RadioButton from '../../../../../components/RadioButton/RadioButton';

class AddCertificationModal extends Component {
  constructor(props) {
    super(props);

    this.state = {
      inputs: {},
      success: false,
      exclusive: false
    };
  }

  componentDidUpdate(prevProps) {
    const { certifications } = this.props;
    if (
      certifications.error &&
      prevProps.certifications.error !== certifications.error
    ) {
      this.handleSubmissionError(certifications.error);
    }
    if (prevProps.certifications.loading && !certifications.loading) {
      this.createCertificationStatus(certifications);
    }
  }

  createCertificationStatus = certificationStatus => {
    if (
      !certificationStatus.error &&
      certificationStatus.data.name !== undefined
    ) {
      this.setState({ success: true });
    }
  };

  handleSubmissionError = error => {
    const { inputs } = this.state;
    if (typeof error === 'object')
      Object.keys(error).forEach(key =>
        inputs[key].setStatus('invalid', error[key][0])
      );
  };

  handleClose = () => {
    const { history } = this.props;
    this.setState({ success: false });
    history.replace('/cadre/certifications');
  };

  handleSubmit = () => {
    const { addCertification } = this.props;
    const {
      inputs: { name, description, duration },
      exclusive
    } = this.state;

    if (!name.isValid()) {
      name.setStatus(
        'invalid',
        'A valid name is required for the certification'
      );
      return false;
    }

    if (!description.isValid()) {
      description.setStatus(
        'invalid',
        'Description is required for the certification'
      );
      return false;
    }

    if (!duration.isValid()) {
      duration.setStatus(
        'invalid',
        'A duration is required for the certification'
      );
      return false;
    }

    const CertificationData = {
      name: name.getValue(),
      description: description.getValue(),
      duration: duration.getValue(),
      exclusive
    };

    addCertification(CertificationData);
    return false;
  };

  saveCertificationButton = () => {
    const { success } = this.state;
    const { certifications } = this.props;
    let button = (
      <button
        type="button"
        className="btn"
        id="saveCertificationButton"
        onClick={this.handleSubmit}
      >
        CREATE
      </button>
    );

    if (success) {
      button = (
        <button
          type="button"
          className="btn"
          id="closeCertificationModal"
          data-dismiss="modal"
          onClick={this.handleClose}
        >
          CLOSE
        </button>
      );
    }
    if (certifications.loading) {
      button = <Loader size="small" />;
    }
    return <div className="modal-footer">{button}</div>;
  };

  handleChange = () => {
    const { exclusive } = this.state;
    this.setState({
      exclusive: !exclusive
    });
  };

  renderRadioButton = exclusive => (
    <RadioButton
      name="exclusive"
      value={exclusive}
      handleChange={this.handleChange}
      placeholder={exclusive}
      options={
        exclusive
          ? 'Certification is Exclusive'
          : 'Certification is not Exclusive'
      }
    />
  );

  renderBody = () => {
    const { exclusive } = this.state;
    return (
      <React.Fragment>
        <FormInputs.TextInput
          parent={this}
          name="name"
          label="Certification Name"
          placeholder="Certification Name"
          testInput={value => value.trim() !== ''}
        />
        <FormInputs.TextInput
          parent={this}
          name="description"
          label="Certification Description"
          type="textarea"
          placeholder="Description"
          testInput={value => value.trim() !== ''}
        />
        <FormInputs.TextInput
          parent={this}
          name="duration"
          label="Duration"
          placeholder="Days"
          testInput={value => value.trim() !== ''}
        />
        {this.renderRadioButton(exclusive)}
      </React.Fragment>
    );
  };

  render() {
    const message = 'New Certification created Successfully';
    const { success } = this.state;

    return (
      <GenericModal
        id="addCertificationModal"
        title="Create Certification"
        footer={this.saveCertificationButton()}
        success={success}
        successMessage={message}
      >
        {this.renderBody()}
      </GenericModal>
    );
  }
}

AddCertificationModal.propTypes = {
  allCertifications: PropTypes.shape({}).isRequired,
  addCertification: PropTypes.shape().isRequired,
  certifications: PropTypes.shape().isRequired,
  history: PropTypes.shape({}).isRequired
};

export default AddCertificationModal;
