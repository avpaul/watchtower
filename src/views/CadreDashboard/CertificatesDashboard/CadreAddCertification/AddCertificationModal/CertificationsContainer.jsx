import { connect } from 'react-redux';
import AddCertificationModal from './AddCertificationModal';
import { addCertification } from '../../../../../redux/actionCreators/cadreAddCertificationAction';

export const mapStateToProps = ({ certifications }) => ({
  certifications
});

export default connect(
  mapStateToProps,
  {
    addCertification
  }
)(AddCertificationModal);
