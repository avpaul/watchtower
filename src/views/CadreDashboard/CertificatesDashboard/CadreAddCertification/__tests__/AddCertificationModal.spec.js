import React from 'react';
import { shallow, mount } from 'enzyme';
import AddCertificationModal from '../AddCertificationModal/AddCertificationModal';
import Loader from '../../../../../components/Loader/Loader';

jest.useFakeTimers();
describe('tests AddCertificationModal', () => {
  const defaultProps = {
    addCertification: jest.fn(),
    certifications: {
      loading: false,
      error: {},
      data: []
    },
    history: {
      replace: jest.fn()
    }
  };

  const setup = (propOverides = {}, shouldMount = false) => {
    const props = { ...defaultProps, ...propOverides };
    const wrapper = shouldMount
      ? mount(
          <div>
            <button
              type="button"
              data-toggle="modal"
              data-target="addCertificationModal"
            >
              Add Certification
            </button>
            <AddCertificationModal {...props} />
          </div>
        )
      : shallow(<AddCertificationModal {...props} />);
    return { props, wrapper };
  };

  beforeEach(() => {
    defaultProps.addCertification = jest.fn();
    jest.runAllTimers();
  });

  it('should Render Components', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('calls handleSubmit successfully', () => {
    const { wrapper } = setup({}, true);
    const buttons = wrapper.find('button');
    buttons.at(0).simulate('click');
  });

  it('should handle errors', () => {
    const modal = mount(<AddCertificationModal {...defaultProps} />);
    defaultProps.addCertification = { name: ['already exists'] };
    const instance = modal.instance();
    jest.spyOn(instance, 'handleSubmissionError');
    modal.instance().handleSubmissionError({ name: ['already exists'] });
    expect(modal.state('inputs').name.state.status).toEqual(5);
  });

  it('should update state on success', () => {
    const modal = mount(<AddCertificationModal {...defaultProps} />);
    defaultProps.addCertification.error = null;
    defaultProps.addCertification.data = { name: 'AWS' };
    const instance = modal.instance();
    jest.spyOn(instance, 'createCertificationStatus');
    modal.instance().createCertificationStatus(defaultProps.addCertification);
    expect(modal.state('success')).toBe(true);
  });

  it('should display loader when loading', () => {
    defaultProps.addCertification.loading = true;
    const modal = mount(<AddCertificationModal {...defaultProps} />);
    const loader = modal.find(Loader);
    expect(loader.length).toEqual(0);
  });
});
