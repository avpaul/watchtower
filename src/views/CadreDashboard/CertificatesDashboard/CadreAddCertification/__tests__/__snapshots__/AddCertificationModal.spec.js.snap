// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests AddCertificationModal should Render Components 1`] = `
Object {
  "props": Object {
    "addCertification": [MockFunction],
    "certifications": Object {
      "data": Array [],
      "error": Object {},
      "loading": false,
    },
    "history": Object {
      "replace": [MockFunction],
    },
  },
  "wrapper": <GenericModal
    footer={
      <div
        className="modal-footer"
      >
        <button
          className="btn"
          id="saveCertificationButton"
          onClick={[Function]}
          type="button"
        >
          CREATE
        </button>
      </div>
    }
    handleClose={[Function]}
    handleSubmit={[Function]}
    id="addCertificationModal"
    submitLoading={false}
    success={false}
    successMessage="New Certification created Successfully"
    title="Create Certification"
  >
    <TextInput
      alertText=""
      className="wt-text-input"
      comment=""
      componentStateKey="inputs"
      defaultStatus={0}
      inputValue=""
      label="Certification Name"
      length={999}
      name="name"
      parent={
        AddCertificationModal {
          "context": Object {},
          "createCertificationStatus": [Function],
          "handleChange": [Function],
          "handleClose": [Function],
          "handleSubmissionError": [Function],
          "handleSubmit": [Function],
          "props": Object {
            "addCertification": [MockFunction],
            "certifications": Object {
              "data": Array [],
              "error": Object {},
              "loading": false,
            },
            "history": Object {
              "replace": [MockFunction],
            },
          },
          "refs": Object {},
          "renderBody": [Function],
          "renderRadioButton": [Function],
          "saveCertificationButton": [Function],
          "setState": [Function],
          "state": Object {
            "exclusive": false,
            "inputs": Object {},
            "success": false,
          },
          "updater": Updater {
            "_callbacks": Array [],
            "_renderer": ReactShallowRenderer {
              "_context": Object {},
              "_didScheduleRenderPhaseUpdate": false,
              "_dispatcher": Object {
                "readContext": [Function],
                "useCallback": [Function],
                "useContext": [Function],
                "useDebugValue": [Function],
                "useEffect": [Function],
                "useImperativeHandle": [Function],
                "useLayoutEffect": [Function],
                "useMemo": [Function],
                "useReducer": [Function],
                "useRef": [Function],
                "useState": [Function],
              },
              "_element": <AddCertificationModal
                addCertification={[MockFunction]}
                certifications={
                  Object {
                    "data": Array [],
                    "error": Object {},
                    "loading": false,
                  }
                }
                history={
                  Object {
                    "replace": [MockFunction],
                  }
                }
              />,
              "_firstWorkInProgressHook": null,
              "_forcedUpdate": false,
              "_instance": [Circular],
              "_isReRender": false,
              "_newState": null,
              "_numberOfReRenders": 0,
              "_renderPhaseUpdates": null,
              "_rendered": <GenericModal
                footer={
                  <div
                    className="modal-footer"
                  >
                    <button
                      className="btn"
                      id="saveCertificationButton"
                      onClick={[Function]}
                      type="button"
                    >
                      CREATE
                    </button>
                  </div>
                }
                handleClose={[Function]}
                handleSubmit={[Function]}
                id="addCertificationModal"
                submitLoading={false}
                success={false}
                successMessage="New Certification created Successfully"
                title="Create Certification"
              >
                <React.Fragment>
                  <TextInput
                    alertText=""
                    className="wt-text-input"
                    comment=""
                    componentStateKey="inputs"
                    defaultStatus={0}
                    inputValue=""
                    label="Certification Name"
                    length={999}
                    name="name"
                    parent={[Circular]}
                    placeholder="Certification Name"
                    testInput={[Function]}
                    type="text"
                  />
                  <TextInput
                    alertText=""
                    className="wt-text-input"
                    comment=""
                    componentStateKey="inputs"
                    defaultStatus={0}
                    inputValue=""
                    label="Certification Description"
                    length={999}
                    name="description"
                    parent={[Circular]}
                    placeholder="Description"
                    testInput={[Function]}
                    type="textarea"
                  />
                  <TextInput
                    alertText=""
                    className="wt-text-input"
                    comment=""
                    componentStateKey="inputs"
                    defaultStatus={0}
                    inputValue=""
                    label="Duration"
                    length={999}
                    name="duration"
                    parent={[Circular]}
                    placeholder="Days"
                    testInput={[Function]}
                    type="text"
                  />
                  <Radio
                    handleChange={[Function]}
                    name="exclusive"
                    options="Certicicate is not Exclusive"
                    placeholder={false}
                    title=""
                    value={false}
                  />
                </React.Fragment>
              </GenericModal>,
              "_rendering": false,
              "_updater": [Circular],
              "_workInProgressHook": null,
            },
          },
        }
      }
      placeholder="Certification Name"
      testInput={[Function]}
      type="text"
    />
    <TextInput
      alertText=""
      className="wt-text-input"
      comment=""
      componentStateKey="inputs"
      defaultStatus={0}
      inputValue=""
      label="Certification Description"
      length={999}
      name="description"
      parent={
        AddCertificationModal {
          "context": Object {},
          "createCertificationStatus": [Function],
          "handleChange": [Function],
          "handleClose": [Function],
          "handleSubmissionError": [Function],
          "handleSubmit": [Function],
          "props": Object {
            "addCertification": [MockFunction],
            "certifications": Object {
              "data": Array [],
              "error": Object {},
              "loading": false,
            },
            "history": Object {
              "replace": [MockFunction],
            },
          },
          "refs": Object {},
          "renderBody": [Function],
          "renderRadioButton": [Function],
          "saveCertificationButton": [Function],
          "setState": [Function],
          "state": Object {
            "exclusive": false,
            "inputs": Object {},
            "success": false,
          },
          "updater": Updater {
            "_callbacks": Array [],
            "_renderer": ReactShallowRenderer {
              "_context": Object {},
              "_didScheduleRenderPhaseUpdate": false,
              "_dispatcher": Object {
                "readContext": [Function],
                "useCallback": [Function],
                "useContext": [Function],
                "useDebugValue": [Function],
                "useEffect": [Function],
                "useImperativeHandle": [Function],
                "useLayoutEffect": [Function],
                "useMemo": [Function],
                "useReducer": [Function],
                "useRef": [Function],
                "useState": [Function],
              },
              "_element": <AddCertificationModal
                addCertification={[MockFunction]}
                certifications={
                  Object {
                    "data": Array [],
                    "error": Object {},
                    "loading": false,
                  }
                }
                history={
                  Object {
                    "replace": [MockFunction],
                  }
                }
              />,
              "_firstWorkInProgressHook": null,
              "_forcedUpdate": false,
              "_instance": [Circular],
              "_isReRender": false,
              "_newState": null,
              "_numberOfReRenders": 0,
              "_renderPhaseUpdates": null,
              "_rendered": <GenericModal
                footer={
                  <div
                    className="modal-footer"
                  >
                    <button
                      className="btn"
                      id="saveCertificationButton"
                      onClick={[Function]}
                      type="button"
                    >
                      CREATE
                    </button>
                  </div>
                }
                handleClose={[Function]}
                handleSubmit={[Function]}
                id="addCertificationModal"
                submitLoading={false}
                success={false}
                successMessage="New Certification created Successfully"
                title="Create Certification"
              >
                <React.Fragment>
                  <TextInput
                    alertText=""
                    className="wt-text-input"
                    comment=""
                    componentStateKey="inputs"
                    defaultStatus={0}
                    inputValue=""
                    label="Certification Name"
                    length={999}
                    name="name"
                    parent={[Circular]}
                    placeholder="Certification Name"
                    testInput={[Function]}
                    type="text"
                  />
                  <TextInput
                    alertText=""
                    className="wt-text-input"
                    comment=""
                    componentStateKey="inputs"
                    defaultStatus={0}
                    inputValue=""
                    label="Certification Description"
                    length={999}
                    name="description"
                    parent={[Circular]}
                    placeholder="Description"
                    testInput={[Function]}
                    type="textarea"
                  />
                  <TextInput
                    alertText=""
                    className="wt-text-input"
                    comment=""
                    componentStateKey="inputs"
                    defaultStatus={0}
                    inputValue=""
                    label="Duration"
                    length={999}
                    name="duration"
                    parent={[Circular]}
                    placeholder="Days"
                    testInput={[Function]}
                    type="text"
                  />
                  <Radio
                    handleChange={[Function]}
                    name="exclusive"
                    options="Certicicate is not Exclusive"
                    placeholder={false}
                    title=""
                    value={false}
                  />
                </React.Fragment>
              </GenericModal>,
              "_rendering": false,
              "_updater": [Circular],
              "_workInProgressHook": null,
            },
          },
        }
      }
      placeholder="Description"
      testInput={[Function]}
      type="textarea"
    />
    <TextInput
      alertText=""
      className="wt-text-input"
      comment=""
      componentStateKey="inputs"
      defaultStatus={0}
      inputValue=""
      label="Duration"
      length={999}
      name="duration"
      parent={
        AddCertificationModal {
          "context": Object {},
          "createCertificationStatus": [Function],
          "handleChange": [Function],
          "handleClose": [Function],
          "handleSubmissionError": [Function],
          "handleSubmit": [Function],
          "props": Object {
            "addCertification": [MockFunction],
            "certifications": Object {
              "data": Array [],
              "error": Object {},
              "loading": false,
            },
            "history": Object {
              "replace": [MockFunction],
            },
          },
          "refs": Object {},
          "renderBody": [Function],
          "renderRadioButton": [Function],
          "saveCertificationButton": [Function],
          "setState": [Function],
          "state": Object {
            "exclusive": false,
            "inputs": Object {},
            "success": false,
          },
          "updater": Updater {
            "_callbacks": Array [],
            "_renderer": ReactShallowRenderer {
              "_context": Object {},
              "_didScheduleRenderPhaseUpdate": false,
              "_dispatcher": Object {
                "readContext": [Function],
                "useCallback": [Function],
                "useContext": [Function],
                "useDebugValue": [Function],
                "useEffect": [Function],
                "useImperativeHandle": [Function],
                "useLayoutEffect": [Function],
                "useMemo": [Function],
                "useReducer": [Function],
                "useRef": [Function],
                "useState": [Function],
              },
              "_element": <AddCertificationModal
                addCertification={[MockFunction]}
                certifications={
                  Object {
                    "data": Array [],
                    "error": Object {},
                    "loading": false,
                  }
                }
                history={
                  Object {
                    "replace": [MockFunction],
                  }
                }
              />,
              "_firstWorkInProgressHook": null,
              "_forcedUpdate": false,
              "_instance": [Circular],
              "_isReRender": false,
              "_newState": null,
              "_numberOfReRenders": 0,
              "_renderPhaseUpdates": null,
              "_rendered": <GenericModal
                footer={
                  <div
                    className="modal-footer"
                  >
                    <button
                      className="btn"
                      id="saveCertificationButton"
                      onClick={[Function]}
                      type="button"
                    >
                      CREATE
                    </button>
                  </div>
                }
                handleClose={[Function]}
                handleSubmit={[Function]}
                id="addCertificationModal"
                submitLoading={false}
                success={false}
                successMessage="New Certification created Successfully"
                title="Create Certification"
              >
                <React.Fragment>
                  <TextInput
                    alertText=""
                    className="wt-text-input"
                    comment=""
                    componentStateKey="inputs"
                    defaultStatus={0}
                    inputValue=""
                    label="Certification Name"
                    length={999}
                    name="name"
                    parent={[Circular]}
                    placeholder="Certification Name"
                    testInput={[Function]}
                    type="text"
                  />
                  <TextInput
                    alertText=""
                    className="wt-text-input"
                    comment=""
                    componentStateKey="inputs"
                    defaultStatus={0}
                    inputValue=""
                    label="Certification Description"
                    length={999}
                    name="description"
                    parent={[Circular]}
                    placeholder="Description"
                    testInput={[Function]}
                    type="textarea"
                  />
                  <TextInput
                    alertText=""
                    className="wt-text-input"
                    comment=""
                    componentStateKey="inputs"
                    defaultStatus={0}
                    inputValue=""
                    label="Duration"
                    length={999}
                    name="duration"
                    parent={[Circular]}
                    placeholder="Days"
                    testInput={[Function]}
                    type="text"
                  />
                  <Radio
                    handleChange={[Function]}
                    name="exclusive"
                    options="Certicicate is not Exclusive"
                    placeholder={false}
                    title=""
                    value={false}
                  />
                </React.Fragment>
              </GenericModal>,
              "_rendering": false,
              "_updater": [Circular],
              "_workInProgressHook": null,
            },
          },
        }
      }
      placeholder="Days"
      testInput={[Function]}
      type="text"
    />
    <Radio
      handleChange={[Function]}
      name="exclusive"
      options="Certicicate is not Exclusive"
      placeholder={false}
      title=""
      value={false}
    />
  </GenericModal>,
}
`;
